### PR TITLE
Do not revert order creation if two orders with the same parameters are created by different users

### PR DESCRIPTION
**Pull request by fedgiac**
One of the main usability pain points of the current smart contract is that two independent users cannot create orders with the exact same parameters. This creates race conditions on order creation between independent users and requires workarounds somewhere in our infrastructure to avoid this issue. The currently planned workaround is [this](https://github.com/cowprotocol/services/issues/804).

These changes should fix the order collision issue. The only remaining case that would cause a transaction to revert unexpectedly on creation is if the same address sends two exactly identical orders (including _real_ validity) one after another.

Drawbacks:
- We need to adapt these changes in our current work. I think most of the work would be in the backend and should be doable. For the frontend, it would change the ABI for order deletion (needs to compute an extra hash) and (if done, not sure) it changes how to get order information onchain (same extra hash).
- More expensive operations. But not too much, < 1000 gas each.
- We need to reaudit the contract. These changes very much affect the security of the whole contract.

*Note*: the PR is a draft insofar that I might improve readability and add further tests. However, the ABI and logic will stay as in this PR so it's worth discussing before spending more time improving it.

The idea behind the changes is that an eth-flow order is added the concept of a "hash" that is different from the CoW Swap one (while still depending on it). This hash includes extra order information like ownership and validity. However, it needs to be presented explicitly on both invalidation and signing, since the information needed to compute it is not part of the order. Also, this extra information needs to be verified against what is available onchain.

### Extra cost

Comparing current `main` with this PR, I notice the following differences:

|            | createOrder | isValidSignature | invalidateOrder | invalidateOrdersIgnoringNotAllowed |
|------------|-------------|------------------|-----------------|------------------------------------|
| main       |       32769 |             1560 |           48327 |                              47441 |
| this PR    |       33002 |             1962 |           48707 |                              48403 |
| difference |        +233 |             +402 |            +380 |                               +962 |

### Test plan

To do. But I already updated all existing unit tests.

### CODE REVIEW

File: src/CoWSwapEthFlow.sol [hunk 1]
```diff
@@ -91,9 +91,11 @@ contract CoWSwapEthFlow is
             revert NotAllowedZeroSellAmount();
         }
 
+        address owner = msg.sender;
```
Comment (nlordell) [inline src/CoWSwapEthFlow.sol:94]: uber-nano-please-ignore-nit: I think `msg.sender` is a `CALLER` opcode which is 2 gas vs using the need to create a stack variable as well as "retrieving" it (`DUP*` or `SWAP*`) which is 3 gas.

Peanuts, but I'm curious IMO doesn't significantly increase readability.

File: src/CoWSwapEthFlow.sol [hunk 2]
```diff
@@ -143,34 +147,42 @@ contract CoWSwapEthFlow is
...
         GPv2Order.Data memory cowSwapOrder = order.toCoWSwapOrder(
             wrappedNativeToken
         );
-        bytes32 orderHash = cowSwapOrder.hash(cowSwapDomainSeparator);
-
-        EthFlowOrder.OnchainData memory orderData = orders[orderHash];
+        bytes32 cowSwapOrderHash = cowSwapOrder.hash(cowSwapDomainSeparator);
+        bytes32 orderHash = EthFlowOrder.hash(cowSwapOrderHash, owner, validTo);
 
+        // We test if the order is valid using user provided values. The validity of these values is confirmad by
```
Comment (nlordell) [inline src/CoWSwapEthFlow.sol:165]:
```suggestion
        // We test if the order is valid using user provided values. The validity of these values is confirmed by
```

File: src/libraries/EthFlowOrder.sol [hunk 3]
```diff
@@ -99,4 +99,12 @@ library EthFlowOrder {
                 GPv2Order.BALANCE_ERC20 // bytes32 buyTokenBalance
             );
     }
+
+    function hash(
+        bytes32 cowSwapOrderHash,
+        address owner,
+        uint32 validTo
+    ) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(cowSwapOrderHash, owner, validTo));
```
Comment (nlordell) [inline src/libraries/EthFlowOrder.sol:108]: Looks like the hash of an order UID 😛.

**Review by nlordell — COMMENTED** [review]
So my understanding is that you use `hash(orderUid)` instead of `hash(order)` to identify orders.

Code looks good, and I don't see any immediate issue.

**Comment by josojo** [comment]
Does this PR break the needed bijection between ethflow orders and technical EIP1271 ethlfow orders on gpv2_contract level?
It seems that two users can place now the same "technical EIP1271 ethflow order" 
(same orderHash, owner=ethflowcontract.address(), validto=inf), right? If this is the case, then the accounting of traded volume on gpv2_contract level will not work out.

Generally, if we solve the challenge on the smart contract level, it's nice - as long as we have the guarantee from Adam that this does not delay the launch.

**Comment by fedgiac** [comment]
Alex is right, the bijection is essential for the contract to work and we can't work around it with some tricks like this. I added a test shat should have passed if the implementation had been good.

Unfortunately I don't see this approach working for us, so we'll go back to our original ideas.

**Pull request closed without merging**